File: high-risk-orders-unactioned/python/flag_high_risk.py
"""Flag high-risk Shopify orders that slipped through without review.

Shopify scores orders for fraud risk, but a HIGH risk order still ships if nobody
looks at it. This lists recent open orders, keeps the ones Shopify rated high risk
(or recommended to cancel) that are not already tagged, and adds a review tag with
tagsAdd so staff catch them before fulfillment. Run on a schedule. Safe to re-run.
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("flag_high_risk")

SHOP = os.environ["SHOPIFY_SHOP"]
TOKEN = os.environ["SHOPIFY_ACCESS_TOKEN"]
API_VERSION = os.environ.get("SHOPIFY_API_VERSION", "2025-01")
ENDPOINT = f"https://{SHOP}/admin/api/{API_VERSION}/graphql.json"
REVIEW_TAG = os.environ.get("REVIEW_TAG", "fraud-review")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

ORDERS_QUERY = """
query($cursor: String) {
  orders(first: 25, after: $cursor, query: "fulfillment_status:unfulfilled AND -status:cancelled") {
    pageInfo { hasNextPage endCursor }
    nodes {
      id name cancelledAt tags
      risk { recommendation assessments { riskLevel } }
    }
  }
}"""

TAGS_ADD = """
mutation($id: ID!, $tags: [String!]!) {
  tagsAdd(id: $id, tags: $tags) {
    node { id }
    userErrors { field message }
  }
}"""


def gql(query, variables=None):
    r = requests.post(
        ENDPOINT,
        json={"query": query, "variables": variables or {}},
        headers={"X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("errors"):
        raise RuntimeError(body["errors"])
    return body["data"]


def is_high_risk(order):
    risk = order.get("risk") or {}
    if risk.get("recommendation") == "CANCEL":
        return True
    return any(a.get("riskLevel") == "HIGH" for a in (risk.get("assessments") or []))


def needs_review(order, review_tag):
    if order.get("cancelledAt"):
        return False
    if review_tag in (order.get("tags") or []):
        return False
    return is_high_risk(order)


def tag_for_review(order_id, review_tag):
    result = gql(TAGS_ADD, {"id": order_id, "tags": [review_tag]})["tagsAdd"]
    if result["userErrors"]:
        raise RuntimeError(result["userErrors"])


def open_orders():
    cursor = None
    while True:
        data = gql(ORDERS_QUERY, {"cursor": cursor})["orders"]
        for node in data["nodes"]:
            yield node
        if not data["pageInfo"]["hasNextPage"]:
            return
        cursor = data["pageInfo"]["endCursor"]


def run():
    flagged = 0
    for order in open_orders():
        if not needs_review(order, REVIEW_TAG):
            continue
        log.warning("Order %s is high risk and unreviewed. %s",
                    order["name"], "would tag" if DRY_RUN else "tagging")
        if not DRY_RUN:
            tag_for_review(order["id"], REVIEW_TAG)
        flagged += 1
    log.info("Done. %d high-risk order(s) %s.", flagged, "to tag" if DRY_RUN else "tagged")


if __name__ == "__main__":
    run()

File: high-risk-orders-unactioned/node/flag-high-risk.js
/**
 * Flag high-risk Shopify orders that slipped through without review.
 *
 * Shopify scores orders for fraud risk, but a HIGH risk order still ships if nobody
 * looks at it. This lists recent open orders, keeps the ones Shopify rated high risk
 * (or recommended to cancel) that are not already tagged, and adds a review tag with
 * tagsAdd so staff catch them before fulfillment. Run on a schedule. Safe to re-run.
 *
 * Guide: https://www.allanninal.dev/shopify/high-risk-orders-unactioned/
 */
import { pathToFileURL } from "node:url";

const SHOP = process.env.SHOPIFY_SHOP || "example.myshopify.com";
const TOKEN = process.env.SHOPIFY_ACCESS_TOKEN || "shpat_dummy";
const API_VERSION = process.env.SHOPIFY_API_VERSION || "2025-01";
const ENDPOINT = `https://${SHOP}/admin/api/${API_VERSION}/graphql.json`;
const REVIEW_TAG = process.env.REVIEW_TAG || "fraud-review";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

export function isHighRisk(order) {
  const risk = order.risk || {};
  if (risk.recommendation === "CANCEL") return true;
  return (risk.assessments || []).some((a) => a.riskLevel === "HIGH");
}

export function needsReview(order, reviewTag) {
  if (order.cancelledAt) return false;
  if ((order.tags || []).includes(reviewTag)) return false;
  return isHighRisk(order);
}

async function gql(query, variables = {}) {
  const res = await fetch(ENDPOINT, {
    method: "POST",
    headers: { "X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Shopify ${res.status}`);
  const body = await res.json();
  if (body.errors) throw new Error(JSON.stringify(body.errors));
  return body.data;
}

const ORDERS_QUERY = `
query($cursor: String) {
  orders(first: 25, after: $cursor, query: "fulfillment_status:unfulfilled AND -status:cancelled") {
    pageInfo { hasNextPage endCursor }
    nodes {
      id name cancelledAt tags
      risk { recommendation assessments { riskLevel } }
    }
  }
}`;

const TAGS_ADD = `
mutation($id: ID!, $tags: [String!]!) {
  tagsAdd(id: $id, tags: $tags) {
    node { id }
    userErrors { field message }
  }
}`;

async function* openOrders() {
  let cursor = null;
  while (true) {
    const data = (await gql(ORDERS_QUERY, { cursor })).orders;
    for (const node of data.nodes) yield node;
    if (!data.pageInfo.hasNextPage) return;
    cursor = data.pageInfo.endCursor;
  }
}

async function tagForReview(orderId, reviewTag) {
  const result = (await gql(TAGS_ADD, { id: orderId, tags: [reviewTag] })).tagsAdd;
  if (result.userErrors.length) throw new Error(JSON.stringify(result.userErrors));
}

export async function run() {
  let flagged = 0;
  for await (const order of openOrders()) {
    if (!needsReview(order, REVIEW_TAG)) continue;
    console.warn(`Order ${order.name} is high risk and unreviewed. ${DRY_RUN ? "would tag" : "tagging"}`);
    if (!DRY_RUN) await tagForReview(order.id, REVIEW_TAG);
    flagged++;
  }
  console.log(`Done. ${flagged} high-risk order(s) ${DRY_RUN ? "to tag" : "tagged"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
